src/realm/util/encrypted_file_mapping.cpp

```diff
@@ -481,14 +486,13 @@ void EncryptedFileMapping::sync() noexcept
 void EncryptedFileMapping::write_barrier(const void* addr, size_t size) noexcept
 {
     REALM_ASSERT(m_access == File::access_ReadWrite);
+    REALM_ASSERT_EX(addr >= m_addr, addr, m_addr);
+    size_t first_accessed_local_page = ((reinterpret_cast<uintptr_t>(addr) - reinterpret_cast<uintptr_t>(m_addr)) >> m_page_shift);
+    size_t num_pages = get_num_pages_for_size(size);
```

Comment:
Yes, good observation. I've fixed it and the whole unit test suite runs faster now which validates your theory that too many pages were refreshed in some cases. It runs in 1m37s now which is an overall improvement of 11 seconds.